### DISCUSSION TITLE
A question about the  pushed image and tags when i  use  the jib  build  strategy

### DISCUSSION THREAD

**Discussion by maodia** [eclipse/jkube#2283] [Q&A]
Hi，i have a question about the  pushed image and tags  when  i  use the jib build  strategy ， so i  need a answer or advice.
We use the Harbor  for docker image registry, and when we use the default build strategy and  the jib build  strategy,  the image and tags display in the Harbor web view is very different.

Here is the image define fragment in my pom file:
```xml
 <image>
      <name>registry.cmread.com/mgsc/demo4k8s-service-consumer</name>
       <build>
           <from>registry.cmread.com/library/openjdk:11-jdk</from>
           <tags>
               <tag>develop</tag>
               <tag>1.0.0</tag>
           </tags>
           <ports>
                 <port>8080</port>
            </ports>
            <assembly>
                 <name>app</name>
            </assembly>
       </build>
 </image>
```

this is a image of one pushed docker image with three tags display in the harber when we use the default build strategy, and it is what we expect:

![0239197caa3502933feb423e3a72bc2](https://github.com/eclipse/jkube/assets/6283382/9bee2cd7-c4e4-4110-b7ab-ad2d5253bab3)

this is a image of three pushed docker image with three tags display in the harber when we use the jib build strategy, and it is not what we expect:
![a30a96920e9cec395b3221ae86bd6ee](https://github.com/eclipse/jkube/assets/6283382/b293a142-dd4d-4187-8e13-a7f6f8402391)

**Comment by rohanKanojia** [comment]
Does your project version containing `-SNAPSHOT`? I remember our plugin's behavior is to add `latest` tag when it's a SNAPSHOT project.

Is unexpected behavior 3 images of 3 tags for you?

**Reply by maodia** [reply]
Yes, out project version containing -SNAPSHOT and it is unexpected for me that 3 images of 3 tags . i think it is normal that  a image with 3 tags.

**Reply by rohanKanojia** [reply]
I copy pasted your configuration into my project and ran `mvn k8s:build k8s:push -Djkube.build.strategy=jib` . You can see image links in DockerHub and Quay:

- DockerHub : https://hub.docker.com/repository/docker/rohankanojia/helidon-starter/general 
- Quay : https://quay.io/repository/rohan.kanojia/helidon-starter?tab=tags

In both repositories I'm able to see them in single repository instead of 3 images of 3 tags.

Do you see this behavior only on harber or on other registries as well?

**Reply by maodia** [reply]
Hi，i browsed  the image links in DockerHub and Quay.  It is true that they are in single repository , and it is the same as i can see  in my Harbor view,but you can see the three tags has three different digest, that means the three tags match three different images.

![image](https://github.com/eclipse/jkube/assets/6283382/f7d7efb3-2f56-43e2-a10e-5a3ded660cd6)

![image](https://github.com/eclipse/jkube/assets/6283382/95e5d017-e697-4cda-a441-c58105e7b97a)

Here is the right sample of alpine:  https://hub.docker.com/_/alpine/tags ,  three tags have the same digest.
![image](https://github.com/eclipse/jkube/assets/6283382/a30bc4a4-d89b-43cd-bc06-125f004274a3)

**Reply by rohanKanojia** [reply]
@maodia : Thanks for clarifying. Yes, it looks like an issue to me. Shall we convert this into GitHub issue?

**Reply by rohanKanojia** [reply]
I think the issue is due to doing separate push request by iterating over tags:

https://github.com/eclipse/jkube/blob/8078c48186e242e734bf3f0c38b3eeba9835b349/jkube-kit/build/service/jib/src/main/java/org/eclipse/jkube/kit/service/jib/JibServiceUtil.java#L147-L151

Instead, we should just pass additional tags in `Containerizer.to(targetImage)` using `.withAdditionalTag` builder method. Loop needs to be moved down here:
https://github.com/eclipse/jkube/blob/8078c48186e242e734bf3f0c38b3eeba9835b349/jkube-kit/build/service/jib/src/main/java/org/eclipse/jkube/kit/service/jib/JibServiceUtil.java#L224

**Reply by maodia** [reply]
@rohanKanojia i think we can convert this into a Github  issue and expect it can be fixed in the future.

**Reply by rohanKanojia** [reply]
@maodia : I've created https://github.com/eclipse/jkube/issues/2293 . Please feel free to add anything in case I missed something.

Do you think it would be possible for you to submit a pull request for this? I can help you out in setting up project and contributing fix.

**Comment by maodia** [comment]
By the way, i am going to ask some questions about the retry mechanism :
(1) Is it suitable for the  jib strategy?
(2) Is it usefull when i pull/push big image from/to the image registry server?
(3) How can i configure the retries element in the xml file? please show me a example.

 
Additional Info:

The network between my pc and the docker image registry server  is  not very well ,  so i always get exceptions when i  build image and push it to the registry with the jib strategy, here is a part of the build logs:


JIB> [======================        ] 74.6% complete > scheduling pushing manifests
JIB> [======================        ] 74.6% complete > scheduling pushing manifests
JIB> [======================        ] 74.7% complete > scheduling pushing manifests
JIB> [31;1mI/O error for image [registry.cmread.com/mgsc/demo4k8s-service-consumer]:[0m                                    
JIB> [31;1m    java.net.SocketTimeoutException[0m                                                                          
JIB> [31;1m    Read timed out[0m                                                                                           
JIB> [31;1mI/O error for image [registry.cmread.com/mgsc/demo4k8s-service-consumer]:[0m                                    
JIB> [31;1m    java.net.SocketTimeoutException[0m                                                                          
JIB> [31;1m    Read timed out[0m                                                                                           
JIB> [=======================       ] 75.8% complete > scheduling pushing manifests
JIB> [=======================       ] 76.9% complete > scheduling pushing manifests
JIB> [31;1mI/O error for image [registry.cmread.com/mgsc/demo4k8s-service-consumer]:[0m                                    
JIB> [31;1m    java.net.SocketTimeoutException[0m                                                                          
JIB> [31;1m    Read timed out[0m                                                                                           
JIB> [=========================     ] 81.8% complete > scheduling pushing manifests
JIB> [31;1mI/O error for image [registry.cmread.com/mgsc/demo4k8s-service-consumer]:[0m                                    
JIB> [31;1m    java.net.SocketException[0m                                                                                 
JIB> [31;1m    Socket closed[0m                                                                                            
JIB> [=========================     ] 81.8% complete > scheduling pushing manifests
Exception occurred while pushing the image: registry.cmread.com/mgsc/demo4k8s-service-consumer:1.0.0, java.net.SocketTimeoutException: Read timed out
Exception occurred while pushing the image: registry.cmread.com/mgsc/demo4k8s-service-consumer
Error when push JIB image [Error when push JIB image]

**Reply by rohanKanojia** [reply]
@maodia : I think you're right, Jib build strategy doesn't seem to respect retry option. I'm not able to find a configuration in JIB docs to enable retry.

They seem to be suggesting using [retry tool](https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#i-am-hitting-docker-hub-rate-limits-how-can-i-configure-registry-mirrors).

**Reply by maodia** [reply]
@rohanKanojia Thanks for your answer,i will try it later.
